src/signalrclient/connection_impl.cpp

```diff
@@ -77,35 +77,12 @@ namespace signalr
             }, m_disconnect_cts.get_token())
             .then([connection](negotiation_response negotiation_response)
             {
-                if (!negotiation_response.try_websockets)
-                {
-                    return pplx::task_from_exception<void>(
-                        std::runtime_error("websockets not supported on the server and there is no fallback transport"));
-                }
-
-                connection->m_connection_token = negotiation_response.connection_token;
-
-                pplx::task_completion_event<void> connect_request_tce;
-
-                auto weak_connection = std::weak_ptr<connection_impl>(connection);
-                auto process_response_callback = [weak_connection, connect_request_tce](const utility::string_t& response)
-                {
-                    auto connection = weak_connection.lock();
-                    if (connection)
+                return connection->start_transport(negotiation_response)
+                    .then([connection, negotiation_response](std::shared_ptr<transport> transport)
                     {
-                        connection->process_response(response, connect_request_tce);
-                    }
-                };
-
-                auto error_callback = [connect_request_tce](const std::exception &e)
-                {
-                    connect_request_tce.set_exception(e);
-                };
-
-                connection->m_transport = connection->m_transport_factory->create_transport(
-                    transport_type::websockets, connection->m_logger, process_response_callback, error_callback);
-
-                return connection->send_connect_request(negotiation_response.connection_token, connect_request_tce);
+                        connection->m_transport = transport;

```

Comment:
While this is correct and saves one atomic operation I would prefer to leave it as is since currently the `transport` parameter is not invalidated and with `std::move()` it will be. This should not matter at the moment but I am afraid of subtle errors introduced by this when maintaining the code. Since this only happens once per connection I don't think this is going to cause any perf issues.